src/interfaces/permissions/request/message.ts

```diff
@@ -0,0 +1,133 @@
+import type { AuthCreateOpts, Authorizable, AuthVerificationResult } from '../../../messages/types';
+import type { JsonPermissionsRequest, PermissionsRequestDescriptor } from './types';
+import type { Scope, Conditions } from '../types';
+
+import { DIDResolver } from '../../../did/did-resolver';
+import { GeneralJwsSigner, GeneralJwsVerifier } from '../../../jose/jws/general';
+import { generateCid, parseCid } from '../../../utils/cid';
+import { isPlainObject } from 'lodash';
+import { Message } from '../../../messages/message';
+import { v4 as uuidv4 } from 'uuid';
+
+type PermissionsRequestOpts = AuthCreateOpts & {
+  conditions?: Conditions;
+  description: string;
+  grantedTo: string;
+  grantedBy: string;
+  objectId?: string;
+  scope: Scope;
+};
+
+export class PermissionsRequest extends Message implements Authorizable {
+  protected message: JsonPermissionsRequest;
+
+  constructor(message: JsonPermissionsRequest) {
+    super(message);
+  }
+
+  get conditions(): Conditions {
+    return this.message.descriptor.conditions;
+  }
+
+  get grantedBy(): string {
+    return this.message.descriptor.grantedBy;
+  }
+
+  get grantedTo(): string {
+    return this.message.descriptor.grantedTo;
+  }
+
+  get description(): string {
+    return this.message.descriptor.description;
+  }
+
+  get scope(): Scope {
+    return this.message.descriptor.scope;
+  }
+
+  static getType(): string {
+    return 'PermissionsRequest';
+  }
+
+  static getInterface(): string {
+    return 'Permissions';
+  }
+
+  static async create(opts: PermissionsRequestOpts): Promise<PermissionsRequest> {
+    const { conditions } = opts;
+    const providedConditions = conditions ? conditions : {};
+    const mergedConditions = { ...DEFAULT_CONDITIONS, ...providedConditions  };
+
+    const descriptor: PermissionsRequestDescriptor = {
+      conditions  : mergedConditions,
+      description : opts.description,
+      grantedTo   : opts.grantedTo,
+      grantedBy   : opts.grantedBy,
+      method      : 'PermissionsRequest',
+      objectId    : opts.objectId ? opts.objectId : uuidv4(),
+      scope       : opts.scope,
+    };
+
+    const descriptorCid = await generateCid(descriptor);
+
+    const authPayload = { descriptorCid: descriptorCid.toString() };
+    const authPayloadStr = JSON.stringify(authPayload);
+    const authPayloadBytes = new TextEncoder().encode(authPayloadStr);
+
+    const signer = await GeneralJwsSigner
+      .create(authPayloadBytes, [opts.signingMaterial]);
```

Comment:
Sanity: I guess we will have to figure out how a custom signer gets passed in the future right?

Reply:
correct